pkg/platform/runtime/setup/setup.go

```diff
@@ -272,24 +269,28 @@ func (s *Setup) updateExecutors(artifacts []artifact.ArtifactID) error {
 	return nil
 }
 
-func (s *Setup) fetchAndInstallArtifacts(installFunc artifactInstaller) error {
+// fetchAndInstallArtifacts needs to return all artifacts needed by the runtime, even if some or
+// all of them were already installed.
+func (s *Setup) fetchAndInstallArtifacts(installFunc artifactInstaller) ([]artifact.ArtifactID, error) {
 	if s.target.InstallFromDir() != nil {
 		return s.fetchAndInstallArtifactsFromDir(installFunc)
 	}
 	return s.fetchAndInstallArtifactsFromRecipe(installFunc)
 }
 
-func (s *Setup) fetchAndInstallArtifactsFromRecipe(installFunc artifactInstaller) error {
+func (s *Setup) fetchAndInstallArtifactsFromRecipe(installFunc artifactInstaller) ([]artifact.ArtifactID, error) {
+	var noArtifacts []artifact.ArtifactID
```

Comment:
Couldn't you just return `nil` rather than having this slice? Same goes for other uses of the `noArtifacts` slice.

Reply:
I didn't know that. Thanks for the suggestion!